src/item_location.cpp

```diff
@@ -478,6 +482,94 @@ class item_location::impl::item_on_vehicle : public item_location::impl
         }
 };
 
+class item_location::impl::item_in_container : public item_location::impl
+{
+    private:
+        item_location container;
+
+        // figures out the index for the item, which is where it is in the total list of contents
+        // note: could be a better way of handling this?
+        int calc_index() const {
+            int idx = 0;
+            for( const item *it : container->contents.all_items_ptr() ) {
+                if( target() == it ) {
+                    return idx;
+                }
+                idx++;
+            }
+            if( idx == 0 ) {
+                return -1;
+            }
+            return idx;
+        }
+    public:
+        item_location parent_item() const override {
+            return container;
+        }
+
+        item_in_container( const item_location &container, item *which ) :
+            impl( which ), container( container ) {}
+
+        void serialize( JsonOut &js ) const override {
+            js.start_object();
+            js.member( "idx", calc_index() );
+            js.member( "type", "in_container" );
+            js.member( "parent", container );
+            js.end_object();
+        }
+
+        item *unpack( int idx ) const override {
+            std::list<const item *> all_items = container->contents.all_items_ptr();
+            auto iter = all_items.begin();
+            std::advance( iter, idx );
```

Comment:
`calc_index()` can return values "-1" (if the container is empty) and `container->contents.all_items_ptr().size()` (if the container is not empty, but the item is not inside the container). That value is stored and later loaded and used here.

And the container item could have changed in the mean time anyway.

In other words, `idx` may not be a valid index anymore, which means calling `std::advance` may increment the iterator beyond the end of the container. Suggestion: add a check for this:
```C++
            if(idx < 0 || static_cast<size_t>(idx) >= all_items.size()) { return nullptr; }
```